### ISSUE TITLE
BlockingIOError: [Errno 11]

### ISSUE THREAD

**Issue by tsmith2** [brinckmann/montepython_public#143]
I am running montepython on a cluster and the mpi runs often (though not always) are killed with the following error:

> Traceback (most recent call last):
>   File "/home/usr/montepython_public_pybird/montepython/io_mp.py", line 511, in lock
>     fcntl.flock(file_obj.fileno(), flags)
> BlockingIOError: [Errno 11] Resource temporarily unavailable

I thought the issue may be that the mpi command was being run on multiple nodes, but this occurs even if I force the job to run on a particular node. 

The admin for the cluster have yet to be able to provide any useful information about how to get around this problem, so any suggestions would be very appreciated.

**Comment by brinckmann** [comment]
That's very strange. I wonder if the chains are not being launched correctly with mpi and different mpi processes try to write to the same chains files.

One thing you can try to do is to launch the job without MPI and instead launch parallel processes/jobs without MPI. As long as you're using Metropolis-Hastings sampling (which is the default) then MontePython doesn't really care whether you use MPI and there's no appreciable gain versus running in parallel. When running in parallel, every chain will be designated as a master and will attempt to compute new covmats. This means a little more time is spent computing covmats, but that time is generally gained back from more efficient adaptation of the covmat and the end result is a wash in efficiency.

Specifically for your case, since you seem to have an I/O issue, I would suggest specifying chain number manually with the flag --chain-number <int> for each parallel process, i.e. passing a succession of 1, 2, 3, 4, ..., etc.

Best,
Thejs

**Comment by flo1984** [comment]
I encountered the same BlockingIOError error on a new iMac machine (not on older machines though). I don't know what the exact reason is but a workaround that worked for me was to keep the cores very busy with another task when starting the mpirun. I used the command stress -c 36 -t 100 to keep all cores running at 100% for 100 seconds (not sure if that command only works for Macs though). After the runs have started successfully and all chain files have been created the problem does not reappear. Best, Florian

**Comment by tsmith2** [comment]
Thanks for the suggestions! It turns out that once I used a more up-to-date openmpi module (openmpi-4.0.2_gcc-8.3.1) the error message was no longer thrown. Though the problem was intermittent, so hopefully it doesn't come back!

**Comment by brinckmann** [comment]
Glad the problem seems to have been resolved! But please check that your chains are being output correctly, i.e. you should have one chains file per mpi process. If that's not the case something is still wrong and it may cause problems down the road.

Best,
Thejs

**Comment by tsmith2** [comment]
It turns out that the problem was not fixed by using the more recent openmpi module. I am now running the jobs without MPI, but would still like to figure out how to use MPI for future runs (which might not use Metropolis-Hastings sampling, for example). I am using slurm to schedule jobs, in case that is relevant.

**Comment by brinckmann** [comment]
I find SLURM and MPI interacts really strangely sometimes, depending on how it was set up. I would suggest finding out how your IT people configured MPI on the cluster. E.g. I found on our cluster I had to use `srun --mpi=pmi2 -n 16 python ...` (`--mpi=pmix` also worked). Unfortunately, it's a little hard for me to provide suggestions on specific cluster setup issues.

Best,
Thejs

**Comment by brinckmann** [comment]
I had one more thought, can you check that mpi4py is correctly installed? MontePython will do strange things if MPI is used, but mpi4py is either not installed or not correctly installed.

Best,
Thejs

**Comment by alexreevesy** [comment]
I had this same issue and it turned out to be due to a faulty installation of mpi4py on the cluster- to fix this I ran: 

source virtual env #sourcing my cluster virtualenv 
pip uninstall mpi4py 
pip install --no-cache-dir mpi4py 

The correct installation of this module can then be checked via ('import mpi4py' was working for me whilst I still had the issue so you need to include the 'MPI' part): 

python 
>>>from mpi4py import MPI 
>>>

Cheers, 

Alex
